consensus/src/lib.rs

```diff
@@ -1,27 +1,1112 @@
+                // Check overflow
+                total_value = (total_value + output.get_value()).ok_or(BlockError::Unknown)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn check_tx_outputs(&self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        for tx in transactions {
+            for _output in tx.get_outputs() {
+                // TODO: Check tx outputs to prevent the overwriting of the transaction
+            }
+        }
+        Ok(())
+    }
+
+    // Connect new block
+    fn connect_tip(
+        &mut self,
+        new_tip_block_index: &BlockIndex,
+        best_block_id: &Option<Id<Block>>,
+    ) -> Result<(), BlockError> {
+        if &self.db_tx.get_best_block_id()? != new_tip_block_index.get_prev_block_id() {
+            return Err(BlockError::Unknown);
+        }
+        let block = self.get_block_from_index(new_tip_block_index)?.expect("Inconsistent DB");
+        let transactions = block.get_transactions();
+
+        if best_block_id.is_some() && !block.is_genesis(self.chain_config) {
+            self.check_block_fee(transactions)?;
+            self.check_tx_inputs(transactions)?;
+        }
+        self.check_tx_outputs(transactions)?;
+        self.connect_transactions(&block)?;
+
+        match &new_tip_block_index.get_prev_block_id() {
+            Some(prev_block_id) => {
+                // To connect a new block we should set-up the next_block_id field of the previous block index
+                let mut prev_block = self
+                    .db_tx
+                    .get_block_index(prev_block_id)?
+                    .expect("Can't get block index. Inconsistent DB");
+                prev_block.set_next_block_id(block.get_id());
+                self.db_tx
+                    .set_block_index(&prev_block)
+                    .expect("Can't set block index. Inconsistent DB");
+            }
+            None => {
+                if best_block_id.is_some() {
+                    panic!("Failed to read block")
+                }
+            }
+        }
+        self.db_tx.set_block_index(new_tip_block_index)?;
+        self.db_tx.set_best_block_id(new_tip_block_index.get_block_id())?;
+        Ok(())
+    }
+
+    fn disconnect_tip(&mut self, block_index: &mut BlockIndex) -> Result<(), BlockError> {
+        assert_eq!(
+            &self.db_tx.get_best_block_id().ok().flatten().expect("Only fails at genesis"),
+            block_index.get_block_id()
+        );
+        let block = self.get_block_from_index(block_index)?.expect("Inconsistent DB");
+        // Disconnect transactions
+        self.disconnect_transactions(block.get_transactions())?;
+        self.db_tx.set_best_block_id(
+            block_index.get_prev_block_id().as_ref().ok_or(BlockError::Unknown)?,
+        )?;
+        // Disconnect block
+        let mut ancestor = self.get_ancestor(block_index.get_block_id())?;
+        ancestor.unset_next_block_id();
+        self.db_tx.set_block_index(&ancestor)?;
```

Comment:
Is the state of the blockchain still valid if transactions are disconnected successfully but disconnecting the block fails for whatever reason?

Reply:
The key idea that when processing a new block, we every time should create a database transaction. If something went wrong, we should make a rollback of database state. In your case, blockchain keep be valid and DB rolled back to state before we met failed block. 

@iljakuklic Rollback called automatically if we don't call `commit` and if we get out of the scope? Or we should explicitly call it?

Reply:
All operations that change the blockchain state are supposed to be done within a transaction. Meaning, no half-changes are possible. Unless the end is reached and a success is returned, nothing changes and changes are reverted.